Pattern Matching Bug

<!-- ISSUE TRACKER ENABLE -->
Aya version: <!-- BEGIN VERSION -->`0.39`<!-- END VERSION -->

<!-- BEGIN FILES -->
```aya
prim I : ISet
prim Path (A : I -> Type) (a : A 0) (b : A 1) : Type
variable A : Type
def infix = (a b : A) : Type => Path (\i => A) a b
def refl {a : A} : a = a => \i => a

def id (A : Type) : A -> A => fn a => a

def test (A : Type) : Fn (B: Type) (b : B) -> B
| a, B => fn b => id B b

open inductive Nat | zero | suc Nat

def eval : test Nat Nat 0 = 0 => refl
```
<!-- END FILES -->

Output:
<details>
<summary>Nat 0 : class org.aya.syntax.core.term.AppTerm</summary>

```
java.lang.AssertionError: Nat 0 : class org.aya.syntax.core.term.AppTerm
	at org.aya.unify.Synthesizer.synthDontNormalize(Synthesizer.java:65)
	at org.aya.unify.DoubleChecker.inherit(DoubleChecker.java:95)
	at org.aya.unify.Unifier.computeReturnType(Unifier.java:161)
	at org.aya.unify.Unifier.doSolveMeta(Unifier.java:64)
	at org.aya.unify.TermComparator.solveMeta(TermComparator.java:92)
	at org.aya.unify.TermComparator.compare(TermComparator.java:193)
	at org.aya.unify.TermComparator.doCompareType(TermComparator.java:527)
	at org.aya.unify.TermComparator.doCompareUntyped(TermComparator.java:323)
```

</details>

--- a/base/src/main/java/org/aya/tyck/StmtTycker.java
+++ b/base/src/main/java/org/aya/tyck/StmtTycker.java
@@ -142,7 +142,7 @@ public record StmtTycker(
                 var rawParams = signature.params();
                 var confluence = new YouTrack(rawParams, tycker, fnDecl.nameSourcePos());
                 var classes = PatClassifier.classify(
-                  patResult.clauses().view().map(Pat.Preclause::pats),
+                  patResult.clauses().view().map(cls -> cls.pats().view()),
                   rawParams.view(), tycker, fnDecl.nameSourcePos());
                 var absurds = patResult.absurdPrefixCount();
                 coreBody.classes = classes.map(cls -> cls.ignoreAbsurd(absurds));
@@ -297,12 +297,12 @@ public record StmtTycker(
       if (lhsResult.hasError()) {
         return;
       }
-      var patWithTypeBound = Pat.collectVariables(lhsResult.freePats().view());
+      var patWithTypeBound = Pat.collectVariables(lhsResult.allPats());
       wellPats = patWithTypeBound.component2();
       tycker.setLocalCtx(lhsResult.localCtx());
       lhsResult.dumpLocalLetTo(ownerBinds, tycker);
       // Here we don't use wellPats but instead a "freePats" because we want them to be bound
-      freeDataCall = new DataCall(dataDef, 0, lhsResult.freePats().map(PatToTerm::visit));
+      freeDataCall = new DataCall(dataDef, 0, lhsResult.allPats().map(PatToTerm::visit).toSeq());
 
       var allTypedBinds = Pat.collectBindings(wellPats.view());
       ownerBinds = patWithTypeBound.component1().toSeq();

--- a/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
+++ b/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
@@ -76,14 +76,14 @@ public final class ClauseTycker implements Problematic, Stateful {
   /// @param paramSubst  substitution for parameter, in the same ordeer as parameter. [Closed] under [#localCtx].
   ///                    Only used by ExprTycker, see [#dumpLocalLetTo]
   /// @param tyckedPats  tycked, free version of the patterns, [Closed] under [#localCtx()]
-  /// @param missingPats patterns that makes [#freePats()] agree with the pusheen signature. Be careful that [#missingPats]
-  ///                    is not necessary to be [Pat.Bind], see [Worker#refinePattern].
+  /// @param missingPats patterns that makes [#allPats()] agree with the pusheen signature. Be careful that [#missingPats]
+  ///                    are not necessarily [Pat.Bind], see [Worker#refinePattern].
   /// @param asSubst     substitution of the `as` patterns, [Closed] under [#localCtx]
-  /// @implNote TL;DR: `freePats.dropLast(unpiParamSize)` is compatible with [#localCtx], [#paramSubst], [#result] and [#body]
-  /// If there are fewer pats than parameters, there will be some pats inserted,
+  /// @implNote TL;DR: [#tyckedPats] is compatible with [#localCtx], [#paramSubst], [#result] and [#body].
+  /// If there are fewer pats than parameters, we insert pats (called [#missingPats]) at the end,
   /// but this will not affect `paramSubst`, and the inserted pat are "ignored" in tycking
   /// of the body, because we check the body against to [#result].
-  /// Then we apply the inserted pats to the body to complete it.
+  /// Then we apply the inserted pats to the body  (essentially using [#allPats()]) to complete it.
   public record LhsResult(
     @NotNull LocalCtx localCtx,
     @NotNull @Closed Term result,
@@ -95,14 +95,11 @@ public final class ClauseTycker implements Problematic, Stateful {
     @NotNull LocalLet asSubst,
     boolean hasError
   ) implements SourceNode {
-    public @NotNull ImmutableSeq<@Closed Pat> freePats() {
-      return tyckedPats.appendedAll(missingPats);
+    public @NotNull SeqView<@Closed Pat> allPats() {
+      return tyckedPats.view().appendedAll(missingPats);
     }
 
-    @Deprecated
-    public int unpiParamSize() {
-      return missingPats.size();
-    }
+    public int userPatSize() { return tyckedPats.size(); }
 
     @Contract(mutates = "param2")
     public void dumpLocalLetTo(@NotNull ImmutableSeq<LocalVar> teleBinds, @NotNull ExprTycker exprTycker) {
@@ -127,7 +124,7 @@ public final class ClauseTycker implements Problematic, Stateful {
       var hasError = lhs.anyMatch(LhsResult::hasError);
       if (!hasError) {
         classes = PatClassifier.classify(
-          lhs.view().map(LhsResult::freePats),
+          lhs.view().map(LhsResult::allPats),
           telescope.view().concat(unpi.params()), parent.exprTycker, overallPos);
         if (clauses.isNotEmpty()) {
           var usages = PatClassifier.firstMatchDomination(clauses, parent, classes);
@@ -135,7 +132,7 @@ public final class ClauseTycker implements Problematic, Stateful {
             if (clauses.get(i).expr.isEmpty()) continue;
             var currentClasses = usages.get(i);
             if (currentClasses.sizeEquals(1)) {
-              var newLhs = refinePattern(lhs.get(i), currentClasses.get(0));
+              var newLhs = refinePattern(lhs.get(i), currentClasses.getAny());
               if (newLhs != null) lhs.set(i, newLhs);
             }
           }
@@ -170,7 +167,7 @@ public final class ClauseTycker implements Problematic, Stateful {
     /// all of these need to be changed accordingly.
     /// This method performs these changes.
     private @Nullable LhsResult refinePattern(LhsResult curLhs, PatClass.Seq<Term, Pat> curCls) {
-      var lets = new PatBinder().apply(curLhs.freePats(), curCls.term());
+      var lets = new PatBinder().apply(curLhs.allPats().toSeq(), curCls.term());
       if (lets.let().let().allMatch((_, j) -> j.wellTyped() instanceof FreeTerm))
         return null;
       var sibling = Objects.requireNonNull(curLhs.localCtx.parent()).derive();
@@ -181,9 +178,9 @@ public final class ClauseTycker implements Problematic, Stateful {
       lets.let().let().forEach(curLhs.asSubst::put);
       return new LhsResult(
         sibling, lets.apply(curLhs.result),
-        newPatterns.dropLast(curLhs.unpiParamSize()),
-        // Well... It is possible if the type has only one constructor, like Unit
-        newPatterns.takeLast(curLhs.unpiParamSize()),
+        newPatterns.take(curLhs.userPatSize()),
+        // We previously assume all the "inserted patterns" are `Pat.Bind`, which is not true after refinement
+        newPatterns.drop(curLhs.userPatSize()),
         curLhs.sourcePos, curLhs.body, paramSubst, curLhs.asSubst, curLhs.hasError);
     }
 
@@ -289,7 +286,7 @@ public final class ClauseTycker implements Problematic, Stateful {
       var bodyExpr = result.body;
       Term wellBody;
       var bindCount = 0;
-      var pats = result.freePats();
+      var pats = result.allPats();
       if (bodyExpr == null) wellBody = null;
       else if (result.hasError) {
         // In case the patterns are malformed, do not check the body
@@ -307,8 +304,8 @@ public final class ClauseTycker implements Problematic, Stateful {
         wellBody = zonker.zonk(wellBody);
 
         // bind all pat bindings
-        var patWithTypeBound = Pat.collectVariables(result.freePats().view());
-        pats = patWithTypeBound.component2();
+        var patWithTypeBound = Pat.collectVariables(result.allPats());
+        pats = patWithTypeBound.component2().view();
         var patBindTele = patWithTypeBound.component1();
 
         bindCount = patBindTele.size();
@@ -318,7 +315,7 @@ public final class ClauseTycker implements Problematic, Stateful {
         wellBody = wellBody.bindTele(patBindTele.view());
       }
 
-      return new Pat.Preclause<>(result.sourcePos, pats, bindCount,
+      return new Pat.Preclause<>(result.sourcePos, pats.toSeq(), bindCount,
         wellBody == null ? null : WithPos.dummy(wellBody));
     }
   }

--- a/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
+++ b/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
@@ -69,14 +69,16 @@ public final class ClauseTycker implements Problematic, Stateful {
     }
   }
 
-  /// @param result     the result according to the pattern tycking, the
-  ///                   [#params] is always empty if the signature result is
-  ///                   [org.aya.tyck.pat.iter.Pusheenable.Const].
-  ///                   The result is always [Closed] under [#localCtx]
-  /// @param paramSubst substitution for parameter, in the same ordeer as parameter. [Closed] under [#localCtx].
-  ///                   Only used by ExprTycker, see [#dumpLocalLetTo]
-  /// @param freePats   free version of the patterns, [Closed] under [#localCtx()]
-  /// @param asSubst    substitution of the `as` patterns, [Closed] under [#localCtx]
+  /// @param result      the result according to the pattern tycking, the
+  ///                    [#params] is always empty if the signature result is
+  ///                    [org.aya.tyck.pat.iter.Pusheenable.Const].
+  ///                    The result is always [Closed] under [#localCtx]
+  /// @param paramSubst  substitution for parameter, in the same ordeer as parameter. [Closed] under [#localCtx].
+  ///                    Only used by ExprTycker, see [#dumpLocalLetTo]
+  /// @param tyckedPats  tycked, free version of the patterns, [Closed] under [#localCtx()]
+  /// @param missingPats patterns that makes [#freePats()] agree with the pusheen signature. Be careful that [#missingPats]
+  ///                    is not necessary to be [Pat.Bind], see [Worker#refinePattern].
+  /// @param asSubst     substitution of the `as` patterns, [Closed] under [#localCtx]
   /// @implNote TL;DR: `freePats.dropLast(unpiParamSize)` is compatible with [#localCtx], [#paramSubst], [#result] and [#body]
   /// If there are fewer pats than parameters, there will be some pats inserted,
   /// but this will not affect `paramSubst`, and the inserted pat are "ignored" in tycking
@@ -84,14 +86,24 @@ public final class ClauseTycker implements Problematic, Stateful {
   /// Then we apply the inserted pats to the body to complete it.
   public record LhsResult(
     @NotNull LocalCtx localCtx,
-    @NotNull @Closed Term result, int unpiParamSize,
-    @NotNull ImmutableSeq<@Closed Pat> freePats,
+    @NotNull @Closed Term result,
+    @NotNull ImmutableSeq<@Closed Pat> tyckedPats,
+    @NotNull ImmutableSeq<@Closed Pat> missingPats,
     @Override @NotNull SourcePos sourcePos,
     @Nullable WithPos<Expr> body,
     @NotNull ImmutableSeq<@Closed Jdg> paramSubst,
     @NotNull LocalLet asSubst,
     boolean hasError
   ) implements SourceNode {
+    public @NotNull ImmutableSeq<@Closed Pat> freePats() {
+      return tyckedPats.appendedAll(missingPats);
+    }
+
+    @Deprecated
+    public int unpiParamSize() {
+      return missingPats.size();
+    }
+
     @Contract(mutates = "param2")
     public void dumpLocalLetTo(@NotNull ImmutableSeq<LocalVar> teleBinds, @NotNull ExprTycker exprTycker) {
       teleBinds.forEachWith(paramSubst, exprTycker.localLet()::put);
@@ -168,7 +180,10 @@ public final class ClauseTycker implements Problematic, Stateful {
       var paramSubst = curLhs.paramSubst.map(jdg -> jdg.map(lets));
       lets.let().let().forEach(curLhs.asSubst::put);
       return new LhsResult(
-        sibling, lets.apply(curLhs.result), curLhs.unpiParamSize, newPatterns,
+        sibling, lets.apply(curLhs.result),
+        newPatterns.dropLast(curLhs.unpiParamSize()),
+        // Well... It is possible if the type has only one constructor, like Unit
+        newPatterns.takeLast(curLhs.unpiParamSize()),
         curLhs.sourcePos, curLhs.body, paramSubst, curLhs.asSubst, curLhs.hasError);
     }
 
@@ -247,13 +262,16 @@ public final class ClauseTycker implements Problematic, Stateful {
       @Closed var instRepi = sigIter.unpiBody().makePi()
         // safe to inst, as paramSubst is closed
         .instTele(patResult.paramSubst().view().map(Jdg::wellTyped));
+      // unpi [instRepi] **at most** [userUnpiSize]
       var instUnpiParam = DepTypeTerm.unpiUnsafe(instRepi, userUnpiSize);
+      // be careful that [freeParam.size() <= userUnpiSize], `==` is not always true.
       var freeParam = AbstractTele.enrich(new AbstractTele.Locns(instUnpiParam.params(), instUnpiParam.body()));
       var missingPats = freeParam.map((x) -> new Pat.Bind(x.ref(), x.type()));
 
-      ImmutableSeq<@Closed Pat> wellTypedPats = patResult.wellTyped().appendedAll(missingPats);
-      return new LhsResult(ctx, instRepi, userUnpiSize,
-        wellTypedPats, clause.sourcePos, patIter.exprBody(),
+      // ImmutableSeq<@Closed Pat> wellTypedPats = patResult.wellTyped().appendedAll(missingPats);
+      return new LhsResult(ctx, instRepi,
+        patResult.wellTyped(), ImmutableSeq.narrow(missingPats),
+        clause.sourcePos, patIter.exprBody(),
         patResult.paramSubst(), patResult.asSubst(), patResult.hasError());
     }
   }
@@ -296,7 +314,7 @@ public final class ClauseTycker implements Problematic, Stateful {
         bindCount = patBindTele.size();
 
         // eta body with inserted patterns
-        wellBody = AppTerm.make(wellBody, pats.view().takeLast(result.unpiParamSize).map(PatToTerm::visit));
+        wellBody = AppTerm.make(wellBody, result.missingPats().view().map(PatToTerm::visit));
         wellBody = wellBody.bindTele(patBindTele.view());
       }
 

--- a/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
+++ b/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
@@ -26,6 +26,7 @@ import org.aya.syntax.telescope.AbstractTele;
 import org.aya.tyck.ExprTycker;
 import org.aya.tyck.TyckState;
 import org.aya.tyck.ctx.LocalLet;
+import org.aya.tyck.error.ClausesProblem;
 import org.aya.tyck.error.PatternProblem;
 import org.aya.tyck.pat.iter.LambdaPusheen;
 import org.aya.tyck.pat.iter.PatternIterator;
@@ -37,6 +38,7 @@ import org.aya.util.position.SourceNode;
 import org.aya.util.position.SourcePos;
 import org.aya.util.position.WithPos;
 import org.aya.util.reporter.Reporter;
+import org.aya.util.tyck.pat.ClassifierUtil;
 import org.aya.util.tyck.pat.PatClass;
 import org.jetbrains.annotations.Contract;
 import org.jetbrains.annotations.NotNull;
@@ -127,17 +129,23 @@ public final class ClauseTycker implements Problematic, Stateful {
           lhs.view().map(LhsResult::allPats),
           telescope.view().concat(unpi.params()), parent.exprTycker, overallPos);
         if (clauses.isNotEmpty()) {
-          var usages = PatClassifier.firstMatchDomination(clauses, parent, classes);
+          var usages = ClassifierUtil.firstMatchDomination(clauses, classes);
           // for the `i`-th clause
           for (int i = 0; i < usages.size(); i++) {
+            var clause = clauses.get(i);
             // skip absurd clauses
-            if (clauses.get(i).expr.isEmpty()) continue;
+            if (clause.expr.isEmpty()) continue;
             var currentClasses = usages.get(i);
-            // if the clause is only reachable for a single leaf in the case tree
-            if (currentClasses.sizeEquals(1)) {
-              // try to refine the patterns
-              var newLhs = refinePattern(lhs.get(i), currentClasses.getAny());
-              if (newLhs != null) lhs.set(i, newLhs);
+            switch (currentClasses.size()) {
+              // if the clause is unreachable
+              case 0 -> parent.fail(new ClausesProblem.FMDomination(i + 1, clause.sourcePos));
+              // if the clause is only reachable for a single leaf in the case tree
+              case 1 -> {
+                // try to refine the patterns
+                var newLhs = refinePattern(lhs.get(i), currentClasses.getAny());
+                if (newLhs != null) lhs.set(i, newLhs);
+              }
+              default -> {}
             }
           }
         }
@@ -161,7 +169,7 @@ public final class ClauseTycker implements Problematic, Stateful {
     /// f x = body2
     ///```
     /// The `x` in the second case is only reachable for input `suc y`,
-    /// and we can realize this by inspecting the result of [PatClassifier#firstMatchDomination].
+    /// and we can realize this by inspecting the result of [ClassifierUtil#firstMatchDomination].
     /// So, we can replace `x` with `suc y` to help computing the result type.
     /// A more realistic motivating example can be found
     /// [here](https://twitter.com/zornsllama/status/1465435870861926400).

--- a/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
+++ b/base/src/main/java/org/aya/tyck/pat/ClauseTycker.java
@@ -128,10 +128,14 @@ public final class ClauseTycker implements Problematic, Stateful {
           telescope.view().concat(unpi.params()), parent.exprTycker, overallPos);
         if (clauses.isNotEmpty()) {
           var usages = PatClassifier.firstMatchDomination(clauses, parent, classes);
+          // for the `i`-th clause
           for (int i = 0; i < usages.size(); i++) {
+            // skip absurd clauses
             if (clauses.get(i).expr.isEmpty()) continue;
             var currentClasses = usages.get(i);
+            // if the clause is only reachable for a single leaf in the case tree
             if (currentClasses.sizeEquals(1)) {
+              // try to refine the patterns
               var newLhs = refinePattern(lhs.get(i), currentClasses.getAny());
               if (newLhs != null) lhs.set(i, newLhs);
             }

--- a/base/src/main/java/org/aya/tyck/pat/PatClassifier.java
+++ b/base/src/main/java/org/aya/tyck/pat/PatClassifier.java
@@ -7,7 +7,6 @@ import kala.collection.SeqView;
 import kala.collection.immutable.ImmutableSeq;
 import kala.collection.immutable.primitive.ImmutableIntSeq;
 import kala.collection.mutable.MutableList;
-import kala.collection.mutable.MutableSeq;
 import kala.control.Result;
 import org.aya.generic.State;
 import org.aya.generic.term.DTKind;
@@ -26,7 +25,6 @@ import org.aya.tyck.tycker.AbstractTycker;
 import org.aya.tyck.tycker.Problematic;
 import org.aya.tyck.tycker.Stateful;
 import org.aya.util.Pair;
-import org.aya.util.position.SourceNode;
 import org.aya.util.position.SourcePos;
 import org.aya.util.reporter.Reporter;
 import org.aya.util.tyck.pat.ClassifierUtil;
@@ -190,15 +188,6 @@ public record PatClassifier(
       case Pat.Bind _ -> new Indexed<>(conTele.view().map(Param::toFreshPat), ix);
       default -> null;
     };
-  }
-
-  /// @return `result[i] : List<T>` means the `i`-th user clause is used by these pat classes
-  public static <T extends PatClass> MutableSeq<MutableList<T>> firstMatchDomination(
-    @NotNull ImmutableSeq<? extends SourceNode> clauses,
-    @NotNull Problematic reporter, @NotNull ImmutableSeq<T> classes
-  ) {
-    return ClassifierUtil.firstMatchDomination(clauses, (pos, i) -> reporter.fail(
-      new ClausesProblem.FMDomination(i, pos)), classes);
   }
 
   private record ConTele(

--- a/base/src/main/java/org/aya/tyck/pat/PatClassifier.java
+++ b/base/src/main/java/org/aya/tyck/pat/PatClassifier.java
@@ -192,6 +192,7 @@ public record PatClassifier(
     };
   }
 
+  /// @return `result[i] : List<T>` means the `i`-th user clause is used by these pat classes
   public static <T extends PatClass> MutableSeq<MutableList<T>> firstMatchDomination(
     @NotNull ImmutableSeq<? extends SourceNode> clauses,
     @NotNull Problematic reporter, @NotNull ImmutableSeq<T> classes

--- a/base/src/main/java/org/aya/tyck/pat/PatClassifier.java
+++ b/base/src/main/java/org/aya/tyck/pat/PatClassifier.java
@@ -66,14 +66,15 @@ public record PatClassifier(
     return subst.appended(term);
   }
 
+  /// @param freePats the second layer of [SeqView] should be friendly for random access
   public static @NotNull ImmutableSeq<PatClass.Seq<Term, Pat>> classify(
-    @NotNull SeqView<ImmutableSeq<Pat>> freePats,
+    @NotNull SeqView<SeqView<Pat>> freePats,
     @NotNull SeqView<Param> telescope, @NotNull AbstractTycker tycker,
     @NotNull SourcePos pos
   ) {
     var classifier = new PatClassifier(tycker, pos);
     var cl = classifier.classifyN(ImmutableSeq.empty(), telescope, freePats
-      .mapIndexed((i, clause) -> new Indexed<>(clause.view(), i))
+      .mapIndexed((i, clause) -> new Indexed<>(clause, i))
       .toSeq(), 4);
     var p = cl.partition(c -> c.cls().isEmpty());
     var missing = p.component1();

--- a/cli-impl/src/test/resources/success/src/Pattern.aya
+++ b/cli-impl/src/test/resources/success/src/Pattern.aya
@@ -49,6 +49,10 @@ module PullRequest1268 {
   def allInOneNoElim {A : Type} (a : A) {B : Type} (b : B) {C : Type} : Fn (D : Type) (d : D) D -> D
   | a, b, D => fn d => id D
 
+  // allInOneNoElim cannot cover the pusheen from body case (but cover the pusheen from signature case)
+  def allInOneNoElim^C (A : Type) (B : Type) : Fn (b : B) B -> B
+  | A => fn B => fn d => id B
+
   // telescope: A a B b C
   // unpi: D d0 -> d1
   @suppress(LocalShadow)

--- a/cli-impl/src/test/resources/success/src/Pattern.aya
+++ b/cli-impl/src/test/resources/success/src/Pattern.aya
@@ -55,3 +55,12 @@ module PullRequest1268 {
   def elims {A : Type} (a : A) (B : Type) (b : B) {C : Type} : Fn (D : Type) D -> D elim a, b
   | a => fn b D => id D
 }
+
+module Issue1380 {
+  def id (A : Type) : A -> A => fn a => a
+
+  def test (A : Type) : Fn (B: Type) (b : B) -> B
+  | a, B => fn b => id B b
+
+  def eval : test Nat Nat 0 = 0 => refl
+}

--- a/tools-kala/src/main/java/org/aya/util/tyck/pat/ClassifierUtil.java
+++ b/tools-kala/src/main/java/org/aya/util/tyck/pat/ClassifierUtil.java
@@ -8,12 +8,10 @@ import kala.collection.mutable.MutableList;
 import kala.collection.mutable.MutableSeq;
 import org.aya.util.Pair;
 import org.aya.util.position.SourceNode;
-import org.aya.util.position.SourcePos;
 import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
 
 import java.util.function.Function;
-import java.util.function.ObjIntConsumer;
 
 public interface ClassifierUtil<Subst, Term, Param, Pat> {
   Param subst(Subst subst, Param param);
@@ -56,17 +54,16 @@ public interface ClassifierUtil<Subst, Term, Param, Pat> {
         .map(args -> args.pair(subclauses)));
   }
 
+  /// @return `result[i] : List<T>` means the `i`-th user clause is used by these pat classes
   static <T extends PatClass> MutableSeq<MutableList<T>> firstMatchDomination(
     @NotNull ImmutableSeq<? extends SourceNode> clauses,
-    @NotNull ObjIntConsumer<SourcePos> report, @NotNull ImmutableSeq<T> classes
+    @NotNull ImmutableSeq<T> classes
   ) {
     // StackOverflow says they're initialized to zero
     var numbers = MutableSeq.fill(clauses.size(), _ -> MutableList.<T>create());
     classes.forEach(results ->
       numbers.get(results.cls().min()).append(results));
     // ^ The minimum is not always the first one
-    for (int i = 0; i < numbers.size(); i++)
-      if (numbers.get(i).isEmpty()) report.accept(clauses.get(i).sourcePos(), i + 1);
     return numbers;
   }
 }

